helper/metricsutil/wrapped_metrics.go

```diff
@@ -0,0 +1,80 @@
+package metricsutil
+
+import (
+	"time"
+
+	metrics "github.com/armon/go-metrics"
+)
+
+// ClusterMetricSink serves as a shim around go-metrics
+// and inserts a "cluster" label.
+//
+// It also provides a mechanism to limit the cardinality of the labels on a gauge
+// (at each reporting interval, which isn't sufficient if there is variability in which
+// labels are the top N) and a backoff mechanism for gauge computation.
+type ClusterMetricSink struct {
+	// ClusterName is either the cluster ID, or a name provided
+	// in the telemetry configuration stanza.
+	ClusterName string
+
+	MaxGaugeCardinality int
+	GaugeInterval       time.Duration
+
+	// Sink is the go-metrics sink to send to
+	Sink metrics.MetricSink
+}
+
+// Convenience alias
+type Label = metrics.Label
+
+func (m *ClusterMetricSink) SetGaugeWithLabels(key []string, val float32, labels []Label) {
+	m.Sink.SetGaugeWithLabels(key, val,
+		append(labels, Label{"cluster", m.ClusterName}))
+}
+
+func (m *ClusterMetricSink) IncrCounterWithLabels(key []string, val float32, labels []Label) {
+	m.Sink.IncrCounterWithLabels(key, val,
+		append(labels, Label{"cluster", m.ClusterName}))
+}
+
+func (m *ClusterMetricSink) AddSampleWithLabels(key []string, val float32, labels []Label) {
+	m.Sink.AddSampleWithLabels(key, val,
+		append(labels, Label{"cluster", m.ClusterName}))
+}
+
+var globalClusterMetrics *ClusterMetricSink
```

Comment:
Every time I introduce a global I always wind up regretting it eventually (and I wish `go-metrics` did not rely on one).  Are you sure this can't just be a field in `Core` instead?

Reply:
It seems to me like we will want to introduce some metrics in places that are not in Core, such as individual secret engines.  The new counts could probably all be handled by adding a member to Core, but then we would end up also passing the wrapper down to TokenStore and ExpirationManager and PassthroughBackend.

On the plus side, having it in core would give a more sensible way of initializing the cluster name. But I don't understand right now how it would get to all the other places.

One of the longer-term goals here is to be able to convert existing go-metrics calls (if we can find a sensible way to do so) so that they can start including the cluster ID too, which pushed me in the direction of more closely matching how go-metrics does things.

I'll give this some more thought.

Reply:
TokenStore and ExpirationManager have their own Core objects already, so I'm not sure why they need to be passed the wrapper.

What about adding something to the system view for use by secrets engines?

Reply:
Good point, they can access it through Core if it's exported.

It seems like the parallel for backends would be to provide it in BackendConfig like the Logger.

Reply:
I would really not like to introduce a metrics sink here and then come back and do it all over again with the event system broker,, but I guess that's probably the best option.

Reply:
Removed global variable; I'll do something with SystemView or BackendConfig when I am adding a real use case.